### ISSUE TITLE
Great work!

### ISSUE THREAD

**Issue by danielhanchen** [c-liuser/secret_puzzles#1]
Actually this is the first submission which actually showed you gave a good effort - great work!

## Task C
On `torch.compile` - I can see you trying very hard to debug which areas are causing issues - you're correct BnB is the main culprit - have you thought about ways to smartly disable them?

You don't need to rewrite whole sections of code as well - but I understand it's because BnB was causing headaches - you could utilize `@torch.compiler.disable` or `@torch._disable_dynamo` to directly disable BnB functions, which can save a lot of time!

I think you forgot (or maybe I misread or missed it) a training run with your changes for the `torch.compile` task - there doesn't seem to be any output in the Colab to verify if graph breaks are still apparent.

Uisng `model = torch.compile(model, fullgraph=False, dynamic=DYNAMIC_SHAPES, options=torch_compile_options)` works for the loss - smart!

```python
if attemped_C:
    C_score = 0
    if uses_flex_attention:
        if dynamic_sequence_length_works: C_score += 3
        else: C_score += 1   ###[+1] Sadly this relies on transformers' Flex Attention
    if no_torch_compile_BnB: C_score -= 2 ###[0] Will ignore this since you tried!
    elif use_part_A: C_score += 1
    elif torch_compile_BnB: C_score += 1  ###[1] Reasonable effort!

    if attention_compiled:
        if excessive_recompilation: C_score -= 3  ###[-1] Sadly the current attention impl will do excessive recompiling. You set flex attention, but compiled the "slower" attention path
        else: C_score += 2
    if mlp_compiled:
        if excessive_recompilation: C_score -= 3
        C_score += 1 ###[+1] I'll give the point!

    if not loss_compiled: C_score -= 1
    if not layernorms_compiled: C_score -= 3

    if max_autotune_triton_matmul:
        if excessive_recompilation: C_score -= 2
        else: C_score += 2

    final_score += C_score
else:
    final_score -= 1

### Final Score = 2 ish for now
```

## Task E
Good attempt! However there is some hardcoding of derivatives, which is fine, but not ideal. Also it's best to transform `nn.CrossEntropyLoss(reduction="mean")` to a direct sum, then divide by N at the end!

I like how you tested it as well!

```python
if attemped_E:
    E_score = 0
    if VRAM_50_percent_reduction: E_score += 2
    if remove_float32_upcast: E_score = 0
    if show_ce_loss_works: E_score += 1
    if show_other_functions_work: E_score += 1
    if hardcoded_gradients: E_score = 0   ###[0] Sadly hardcoding
    if allows_dynamic_chunk_sizes: E_score += 1
    if llama_1B_training_loss_matches: E_score += 1
    else: E_score = 0
    if GRPO_memory_efficient_linear_works: E_score += 4
    final_score += E_score
else:
    final_score += 0

### Final Score = 0 for now, however + 2 for effort!
```

## Task B
I haven't ran it in Kaggle yet, but I took a skim - I think this is a mainly FSDP1 impl correct? FSDP2 is not yet supported by accelerate.

However I like the notebook and all the code - great work!

For now I'll give + 4 points for effort :)

## Other stuff
In general, fantastic attempt - 8 points for now - the tasks were more of trying to see if you put in effort and whether you took responsibility for testing and stuff. When are looking for an internship during your uni studies? Which terms and stuff?

**Comment by c-liuser** [comment]
@danielhanchen 

First of all, thank you for taking the time to review my submission so thoroughly, I'll continue refining it based on your feedback.

As for the internship, I’m free from now until late March. After that, I’ll be abroad for an exchange program from April 1 to August 20 (considered my Spring semester). I'm sure I would have some time to dedicate to an internship during this period, but it would likely be complicated and I know you mentioned U.S. remote preferred. Once I return, however, a fall semester internship would be a great fit for me. I will have 2 semesters remaining following this Spring semester, for context.

Let me know if you have any questions about this.

**Comment by danielhanchen** [comment]
Ok cool! I'll get back to you on internships - it's probably going to be a bit hard on timing! Definitely good effort though!

**Comment by c-liuser** [comment]
Sounds good! I figured it might be difficult timing, but I hope we can work something out if you think I'd be a good fit.

I reworked E and I believe now it doesn't have hardcoded derivatives and uses autograd instead. It improved the VRAM reduction very slightly as well. Please take a look if you have a chance.

I'm still working through B and C, having some trouble. I provided 2 solutions for C, the first of which has a few graph breaks and was a more simple attempt. The second crashes my colab session for whatever reason and I can't figure out why.

I'm also working on properly implementing FSDP2 and it'll take some more research before I can get it to work properly. A lot of the logic I previously used doesn't directly translate so I'm breaking it down piece by piece at the moment.

**Comment by c-liuser** [comment]
@danielhanchen 

I'm still working through the FSDP2 task. I'm trying to handle OOM errors at the moment.

Do you have an update for me regarding the internship?

**Comment by danielhanchen** [comment]
Hey! Apologies on the delay! It's probably going to be extremely tough sorry on the internship - appreciate the effort and good luck on your endeavors!
